File: backend/src/main/java/com/lsoft/aguapotable/service/MedidorService.java
package com.lsoft.aguapotable.service;

import com.lsoft.aguapotable.domain.entity.Medidor;
import com.lsoft.aguapotable.domain.entity.Suscriptor;
import com.lsoft.aguapotable.domain.enums.EstadoMedidor;
import com.lsoft.aguapotable.domain.repository.MedidorRepository;
import com.lsoft.aguapotable.domain.repository.SuscriptorRepository;
import com.lsoft.aguapotable.dto.request.MedidorRequest;
import com.lsoft.aguapotable.exception.BusinessException;
import com.lsoft.aguapotable.exception.ResourceNotFoundException;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.util.List;

@Service
@RequiredArgsConstructor
public class MedidorService {

    private final MedidorRepository medidorRepository;
    private final SuscriptorRepository suscriptorRepository;

    public List<Medidor> getPorSuscriptor(Long suscriptorId) {
        return medidorRepository.findBySuscriptorId(suscriptorId);
    }

    public List<Medidor> getPorComunidadYEstado(Long comunidadId, EstadoMedidor estado) {
        return medidorRepository.findBySuscriptorComunidadIdAndEstado(comunidadId, estado);
    }

    public Medidor obtenerPorId(Long id, Long comunidadId) {
        Medidor medidor = medidorRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Medidor", id));
        if (!medidor.getSuscriptor().getComunidad().getId().equals(comunidadId)) {
            throw new ResourceNotFoundException("Medidor", id);
        }
        return medidor;
    }

    @Transactional
    public Medidor crear(MedidorRequest request, Long comunidadId) {
        Suscriptor suscriptor = suscriptorRepository.findById(request.getSuscriptorId())
                .orElseThrow(() -> new ResourceNotFoundException("Suscriptor", request.getSuscriptorId()));

        if (!suscriptor.getComunidad().getId().equals(comunidadId)) {
            throw new ResourceNotFoundException("Suscriptor", request.getSuscriptorId());
        }

        if (medidorRepository.existsByNumeroSerie(request.getNumeroSerie())) {
            throw new BusinessException("Ya existe un medidor con serie: " + request.getNumeroSerie());
        }

        Medidor medidor = Medidor.builder()
                .numeroSerie(request.getNumeroSerie())
                .marca(request.getMarca())
                .diametro(request.getDiametro())
                .lecturaInicial(request.getLecturaInicial() != null ? request.getLecturaInicial() : 0.0)
                .estado(request.getEstado() != null ? request.getEstado() : EstadoMedidor.ACTIVO)
                .fechaInstalacion(request.getFechaInstalacion() != null ? request.getFechaInstalacion() : LocalDate.now())
                .suscriptor(suscriptor)
                .build();

        return medidorRepository.save(medidor);
    }

    @Transactional
    public Medidor actualizar(Long id, MedidorRequest request, Long comunidadId) {
        Medidor medidor = obtenerPorId(id, comunidadId);

        medidor.setNumeroSerie(request.getNumeroSerie());
        medidor.setMarca(request.getMarca());
        medidor.setDiametro(request.getDiametro());
        medidor.setEstado(request.getEstado());

        if (request.getSuscriptorId() != null &&
                !request.getSuscriptorId().equals(medidor.getSuscriptor().getId())) {
            Suscriptor nuevoSuscriptor = suscriptorRepository.findById(request.getSuscriptorId())
                    .orElseThrow(() -> new ResourceNotFoundException("Suscriptor", request.getSuscriptorId()));
            medidor.setSuscriptor(nuevoSuscriptor);
        }

        return medidorRepository.save(medidor);
    }

    @Transactional
    public Medidor darDeBaja(Long id, Long comunidadId) {
        Medidor medidor = obtenerPorId(id, comunidadId);
        medidor.setEstado(EstadoMedidor.RETIRADO);
        medidor.setFechaBaja(LocalDate.now());
        return medidorRepository.save(medidor);
    }
}

File: backend/src/medidor/medidor.service.ts
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { Medidor, Suscriptor } from '../common/entities';
import { EstadoMedidor } from '../common/enums';
import { MedidorRequest } from './dto/medidor.dto';
import {
  BusinessException,
  ResourceNotFoundException,
} from '../common/exceptions';

@Injectable()
export class MedidorService {
  constructor(
    @InjectRepository(Medidor)
    private readonly repo: Repository<Medidor>,
    @InjectRepository(Suscriptor)
    private readonly suscriptorRepo: Repository<Suscriptor>,
  ) {}

  async getPorSuscriptor(suscriptorId: number): Promise<Medidor[]> {
    return this.repo.find({ where: { suscriptorId } });
  }

  async getPorComunidadYEstado(
    comunidadId: number,
    estado: EstadoMedidor,
  ): Promise<Medidor[]> {
    return this.repo
      .createQueryBuilder('m')
      .innerJoinAndSelect('m.suscriptor', 's')
      .where('s.comunidad_id = :comunidadId', { comunidadId })
      .andWhere('m.estado = :estado', { estado })
      .getMany();
  }

  async obtenerPorId(id: number, comunidadId: number): Promise<Medidor> {
    const medidor = await this.repo.findOne({
      where: { id },
      relations: ['suscriptor', 'suscriptor.comunidad'],
    });
    if (!medidor) throw new ResourceNotFoundException('Medidor', id);
    if (medidor.suscriptor.comunidadId !== Number(comunidadId)) {
      throw new ResourceNotFoundException('Medidor', id);
    }
    return medidor;
  }

  async crear(dto: MedidorRequest, comunidadId: number): Promise<Medidor> {
    const suscriptor = await this.suscriptorRepo.findOne({
      where: { id: dto.suscriptorId },
    });
    if (!suscriptor)
      throw new ResourceNotFoundException('Suscriptor', dto.suscriptorId);
    if (suscriptor.comunidadId !== Number(comunidadId)) {
      throw new ResourceNotFoundException('Suscriptor', dto.suscriptorId);
    }

    const existe = await this.repo.findOne({
      where: { numeroSerie: dto.numeroSerie },
    });
    if (existe) {
      throw new BusinessException(
        'Ya existe un medidor con serie: ' + dto.numeroSerie,
      );
    }

    const medidor = this.repo.create({
      numeroSerie: dto.numeroSerie,
      marca: dto.marca ?? null,
      diametro: dto.diametro ?? null,
      lecturaInicial: dto.lecturaInicial ?? 0,
      estado: dto.estado ?? EstadoMedidor.ACTIVO,
      fechaInstalacion:
        dto.fechaInstalacion ?? new Date().toISOString().split('T')[0],
      suscriptorId: dto.suscriptorId,
    });
    return this.repo.save(medidor);
  }

  async actualizar(
    id: number,
    dto: MedidorRequest,
    comunidadId: number,
  ): Promise<Medidor> {
    const medidor = await this.obtenerPorId(id, comunidadId);
    medidor.numeroSerie = dto.numeroSerie;
    medidor.marca = dto.marca ?? null;
    medidor.diametro = dto.diametro ?? null;
    medidor.estado = dto.estado ?? medidor.estado;

    if (dto.suscriptorId && dto.suscriptorId !== medidor.suscriptorId) {
      const nuevoSuscriptor = await this.suscriptorRepo.findOne({
        where: { id: dto.suscriptorId },
      });
      if (!nuevoSuscriptor)
        throw new ResourceNotFoundException('Suscriptor', dto.suscriptorId);
      medidor.suscriptorId = dto.suscriptorId;
    }

    return this.repo.save(medidor);
  }

  async darDeBaja(id: number, comunidadId: number): Promise<Medidor> {
    const medidor = await this.obtenerPorId(id, comunidadId);
    medidor.estado = EstadoMedidor.RETIRADO;
    medidor.fechaBaja = new Date().toISOString().split('T')[0];
    return this.repo.save(medidor);
  }
}
